pkg/option/config.go

```diff
@@ -15,71 +15,123 @@
+	// disabled.
+	AllowLocalhostPolicy = "policy"
+
+	// ModePreFilterNative for loading progs with xdpdrv
+	ModePreFilterNative = "native"
+
+	// ModePreFilterGeneric for loading progs with xdpgeneric
+	ModePreFilterGeneric = "generic"
+)
+
+// Config is the configuration used by Daemon.
+type Config struct {
+	BpfDir          string     // BPF template files directory
+	LibDir          string     // Cilium library files directory
+	RunDir          string     // Cilium runtime directory
+	NAT46Prefix     *net.IPNet // NAT46 IPv6 Prefix
+	Device          string     // Receive device
+	DevicePreFilter string     // XDP device
+	ModePreFilter   string     // XDP mode, values: { native | generic }
+	HostV4Addr      net.IP     // Host v4 address of the snooping device
+	HostV6Addr      net.IP     // Host v6 address of the snooping device
+	IPv4Disabled    bool       // Disable IPv4 allocation
+	LBInterface     string     // Set with name of the interface to loadbalance packets from
+
+	Tunnel string // Tunnel mode
+
+	DryMode       bool // Do not create BPF maps, devices, ..
+	RestoreState  bool // RestoreState restores the state from previous running daemons.
+	KeepConfig    bool // Keep configuration of existing endpoints when starting up.
+	KeepTemplates bool // Do not overwrite the template files
+
+	// AllowLocalhost defines when to allows the local stack to local endpoints
+	// values: { auto | always | policy }
+	AllowLocalhost string
+
+	// StateDir is the directory where runtime state of endpoints is stored
+	StateDir string
+
+	// Options changeable at runtime
+	Opts *BoolOptions
 
-// NamedMapOptions is a MapOptions struct with a configuration name.
-// This struct is useful to keep reference to the assigned
-// field name in the internal configuration struct.
-type NamedMapOptions struct {
-	name string
-	MapOptions
+	// Mutex for serializing configuration updates to the daemon.
+	ConfigPatchMutex lock.RWMutex
+
+	// Monitor contains the configuration for the node monitor.
+	Monitor *models.MonitorStatus
+
+	// AccessLog is the path to the access log of supported L7 requests observed.
+	AccessLog string
+
+	// AgentLabels contains additional labels to identify this agent in monitor events.
+	AgentLabels []string
 }
 
-// NewNamedMapOptions creates a reference to a new NamedMapOpts struct.
-func NewNamedMapOptions(name string, values *map[string]string, validator Validator) *NamedMapOptions {
-	return &NamedMapOptions{
-		name:       name,
-		MapOptions: *NewMapOpts(*values, validator),
-	}
+var (
+	config *Config
+	once   sync.Once
+)
+
+// GetConfig returns the global configuration set in flags
+func GetConfig() *Config {
+	once.Do(func() {
+		if config == nil {
+			config = &Config{
+				Opts:    NewBoolOptions(&daemonLibrary),
+				Monitor: &models.MonitorStatus{Cpus: int64(runtime.NumCPU()), Npages: 64, Pagesize: int64(os.Getpagesize()), Lost: 0, Unknown: 0},
+			}
+		}
+	})
+	return config
 }
```

Comment:
Couldn't GetConfig be replaced by an init? There is an order of operations that is important to get the options setup, but using the once vs init would have the same problem: what if we call GetConfig() before the options are initialised (which we do in main, I think), but using init would have a more consistent outcome. If we did that, we can then make `config` a global, to better reflect that it is the exact same object everywhere.

Reply:
In fact, is there a reason we can't just initialise it directly, where it's declared?